### PR TITLE
Pass user to isCompleted() via _resolve_user() (#1968)

### LINKED ISSUE #1968: ProgramModuleObj.isCompleted() should take a user
Right now `ProgramModuleObj.isCompleted()` just takes `self` as a parameter.  Nearly all callers of it already have the current user available, and nearly all implementations of it then end up calling `get_current_request().user` to get the current user.  This is ugly; we should pass the user as an argument.

### PR DESCRIPTION
**Pull request by imran-techvoyager**
## Description

This PR refactors `ProgramModuleObj.isCompleted()` to accept an explicit `user` parameter and systematically propagates that change across all module handlers and call sites.

Historically, `isCompleted()` implicitly resolved the user from `self.user` or `get_current_request()`, which tightly coupled module logic to request context and introduced subtle ambiguity in multi-user flows and template execution. This change makes user context explicit while preserving backward compatibility.

### Key Changes

* Introduced `_resolve_user(self, user)` in `modules/base.py`.

  * Resolution precedence:

    1. Explicit argument
    2. `self.user` (set via `Program.getModules`)
    3. `get_current_request().user` (template fallback)
* Updated all `isCompleted()` overrides across module handlers to:

  * Accept `user=None`
  * Call `_resolve_user(user)` internally
* Updated all call sites to pass `request.user` where appropriate.
* Updated sorting logic in `Program.getModules()` to use `mod.isCompleted(user)`.
* Added targeted tests for user resolution behavior.
* Preserved legacy behavior for template contexts and transitional call sites.

This refactor reduces hidden state coupling and prepares the module system for safer multi-user and non-request-bound execution contexts.

---

## Related Issue

Closes #1968

---

## Type of Change

* [ ] Bug fix
* [ ] New feature
* [ ] Breaking change
* [ ] Documentation update
* [x] Refactor / cleanup
* [ ] CI/CD or build change

---

## Breaking Change Considerations

This change modifies the public signature of `isCompleted()` across all modules to accept `user=None`. However:

* A default value is preserved.
* `_resolve_user()` maintains prior implicit behavior.
* Existing template calls continue to function.

Functionally, behavior should remain unchanged. The primary impact is architectural correctness and improved clarity of user scoping.

Given the wide surface area (40 files), this PR should be considered medium risk despite preserving behavior.

---

## Risk Areas Reviewed

Manual smoke testing performed for:

* Student registration flow
* Teacher registration flow
* Credit card payment validation
* Required-module gating logic
* Self-checkin flow

All updated module completion checks now consistently use the explicit user context.

Particular attention was paid to:

* `findModule` routing logic
* Required-module completion guards
* Sorting logic in `getModules`
* Payment flow checks (`CreditCardModule_*`)
* Core confirmation flow (`StudentRegCore`, `StudentRegConfirm`)

---

## Testing

* [x] Existing tests pass
* [x] New tests added
* [x] Manually verified

New tests:

* Added resolution tests to ensure `_resolve_user()` precedence works correctly.
* Verified no cross-user leakage in module completion checks.

---

## Checklist

* [x] Self-reviewed the code
* [ ] Updated documentation (not required)
* [x] No new warnings or errors introduced

### CODE REVIEW

**Comment by imran-techvoyager** [comment]
Hi @willgearty this one is ready for review when you get a chance.
All checks are passing and tests are green.

Would appreciate your thoughts. Thanks!

File: esp/esp/program/modules/tests/resolve_user.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        """Explicit argument must win over self.user."""
+        self.moduleobj.user = self.students[0]
+        resolved = self.moduleobj._resolve_user(self.students[1])
+        self.assertEqual(resolved, self.students[1],
+                         "_resolve_user should return the explicit argument, not self.user")
+
+    def test_self_user_takes_priority_over_request(self):
+        """self.user must win over get_current_request().user."""
+        get_current_request().user = self.students[1]
+        self.moduleobj.user = self.students[0]
```
[the remark this answers is not in the document]
Comment (imran-techvoyager) [reply esp/esp/program/modules/tests/resolve_user.py:47]: Thanks for the catch, this is now addressed.

The test now creates a controlled RequestFactory request in setUp, pushes it into the thread-local explicitly, and clears it in tearDown to avoid order-dependent behavior.

This ensures _resolve_user() is exercised deterministically without relying on implicit request state.

File: esp/esp/program/modules/base.py [hunk 2]
```diff
@@ -336,7 +336,19 @@ def isOnSiteFeatured(self):
...
+        """Return the user to use for per-user module logic.
+
+        Precedence: explicit argument > self.user (set by getModules) >
+        get_current_request().user (fallback for template calls).
+        """
+        if user is not None:
+            return user
+        if hasattr(self, 'user'):
+            return self.user
+        return get_current_request().user
```
[the remark this answers is not in the document]
Comment (imran-techvoyager) [reply esp/esp/program/modules/base.py:354]: Good point regarding clearer failure semantics.

For this PR, I intentionally preserved the previous behavior (including the implicit AttributeError when no request is available), since the scope here is refactoring user resolution without altering runtime behavior.

If you’d prefer explicit errors in non-request contexts, I’m happy to follow up with a separate change that addresses that more deliberately.

**Comment by imran-techvoyager** [comment]
@willgearty , this one is ready for review when you have time.

Given the surface area (explicit `user` propagation across modules), I’d especially appreciate a second set of eyes on the `_resolve_user()` precedence and any edge cases you think I may have missed. All tests are passing.
